### ISSUE TITLE
DivideByZeroException is thrown during diff commits

### ISSUE THREAD

**Issue by DaveSenn** [libgit2/libgit2sharp#979]
Sometimes a get a DivideByZeroException exception when I try to diff two commits.
I’m using LibGit2Sharp (Nuget-Package version 0.21.0.176) in a .Net 4.5.2 application.

My Code:

```csharp
var parentCommit = commit.Parents.FirstOrDefault();
if ( gitCommit.Changes == null )
	gitCommit.Changes = new Collection<GitChange>();

if ( parentCommit != null )
{
	var changes = repository.Diff.Compare<TreeChanges>( parentCommit.Tree, commit.Tree );
	changes.Take( Parameters.MaxChangesToTrack )
		   .ForEach( x =>
		   {
			   var gitChange = new GitChange
			   {
				   Commit = gitCommit,
				   Sha = x.Oid.Sha,
				   OldSha = x.OldOid.Sha,
				   Path = x.Path,
				   OldPath = x.OldPath,
				   Mode = x.Mode.ToGitMode(),
				   OldMode = x.OldMode.ToGitMode(),
				   ChangeKind = x.Status.ToGitChangeKind()
			   };
			   gitCommit.Changes.Add( gitChange );
		   } );
}
else
	commit.Tree.ForEach( x =>
	{
		var gitChange = new GitChange
		{
			Commit = gitCommit,
			Sha = x.Target.Sha,
			OldSha = String.Empty,
			Path = x.Path,
			OldPath = String.Empty,
			Mode = x.Mode.ToGitMode(),
			OldMode = GitMode.Nonexistent,
			ChangeKind = GitChangeKind.Added
		};
		gitCommit.Changes.Add( gitChange );
	} );
```

Exception:
```csharp
System.DivideByZeroException: Attempted to divide by zero.
   at LibGit2Sharp.Core.NativeMethods.git_diff_find_similar(DiffSafeHandle diff, GitDiffFindOptions options)
   at LibGit2Sharp.Core.Proxy.git_diff_find_similar(DiffSafeHandle diff, GitDiffFindOptions options)
   at LibGit2Sharp.Diff.DetectRenames(DiffSafeHandle diffList, CompareOptions compareOptions)
   at LibGit2Sharp.Diff.BuildDiffList(ObjectId oldTreeId, ObjectId newTreeId, TreeComparisonHandleRetriever comparisonHandleRetriever, DiffModifiers diffOptions, IEnumerable`1 paths, ExplicitPathsOptions explicitPathsOptions, CompareOptions compareOptions)
   at LibGit2Sharp.Diff.Compare[T](Tree oldTree, Tree newTree, IEnumerable`1 paths, ExplicitPathsOptions explicitPathsOptions, CompareOptions compareOptions)
   at GitTool.RepositoryService`1.AddGitChanges(Commit commit, Repository repository, GitCommit gitCommit) in c:\GitTool\RepositoryService.cs:line 284
   at GitTool.RepositoryService`1.<>c__DisplayClasse.<AddCommits>b__b(NewGitCommit x) in c:\GitTool\RepositoryService.cs:line 194
   at PortableExtensions.IEnumerableTEx.ForEach[T](IEnumerable`1 enumerable, Action`1 action)
   at GitTool.RepositoryService`1.AddCommits(IEnumerable`1 newCommits, Repository repository, Repository gitRepository) in c:\GitTool\RepositoryService.cs:line 158
   at GitTool.GitLogReader.FindNewCommits(String repositoryName, String repositoryDirectory) in c:\GitTool\GitLogReader.cs:line 115
   at GitTool.GitManager.<>c__DisplayClass7.<WorkOffNotifications>b__5() in c:\GitTool\GitManager.cs:line 146    
   at LibGit2Sharp.Core.NativeMethods.git_diff_find_similar(DiffSafeHandle diff, GitDiffFindOptions options)
   at LibGit2Sharp.Core.Proxy.git_diff_find_similar(DiffSafeHandle diff, GitDiffFindOptions options)
   at LibGit2Sharp.Diff.DetectRenames(DiffSafeHandle diffList, CompareOptions compareOptions)
   at LibGit2Sharp.Diff.BuildDiffList(ObjectId oldTreeId, ObjectId newTreeId, TreeComparisonHandleRetriever comparisonHandleRetriever, DiffModifiers diffOptions, IEnumerable`1 paths, ExplicitPathsOptions explicitPathsOptions, CompareOptions compareOptions)
   at LibGit2Sharp.Diff.Compare[T](Tree oldTree, Tree newTree, IEnumerable`1 paths, ExplicitPathsOptions explicitPathsOptions, CompareOptions compareOptions)
   at GitTool.RepositoryService`1.AddGitChanges(Commit commit, Repository repository, GitCommit gitCommit) in c:\GitTool\RepositoryService.cs:line 284
   at GitTool.RepositoryService`1.<>c__DisplayClasse.<AddCommits>b__b(NewGitCommit x) in c:\GitTool\RepositoryService.cs:line 194
   at PortableExtensions.IEnumerableTEx.ForEach[T](IEnumerable`1 enumerable, Action`1 action)
   at GitTool.RepositoryService`1.AddCommits(IEnumerable`1 newCommits, Repository repository, Repository gitRepository) in c:\GitTool\RepositoryService.cs:line 158
   at GitTool.GitLogReader.FindNewCommits(String repositoryName, String repositoryDirectory) in c:\GitTool\GitLogReader.cs:line 115
   at GitTool.GitManager.<>c__DisplayClass7.<WorkOffNotifications>b__5() in c:\GitTool\GitManager.cs:line 146
```

Am I doing something wrong?

**Comment by nulltoken** [comment]
>Am I doing something wrong?

Doesn't look like so.

Can you always reproduce this? Meaning, when it fails performing a diff between two commits, it'll keep on failing if you try again with the same two commits?

In order to troubleshoot this in an efficient manner, we'd need some kind of repro case against a public repository. Any chance you could make this happen?

**Comment by DaveSenn** [comment]
Yes it’s reproducible.
Unfortunately I’m not allowed to publish any part of this repository to a public repository. 

But I found a trace…
The commit, which causes the exception, has two parents. (Commit has merged master and another branch) While all other commits only have one.

I’ve attached two images showing the log-tree.
Hop it helps… 

![gitk](https://cloud.githubusercontent.com/assets/7510324/6347957/5426215a-bc1c-11e4-9329-40252a647d0b.PNG)
![tree](https://cloud.githubusercontent.com/assets/7510324/6347958/5426a74c-bc1c-11e4-808b-208d26e18513.PNG)

**Comment by DaveSenn** [comment]
It works when I compare the commit and the last parent (instead of the first parent).
```csharp
//var parentCommit = commit.Parents.FirstOrDefault();
var parentCommit = commit.Parents.LastOrDefault();
```

**Comment by nulltoken** [comment]
@DaveSenn Would you be able to find an existing public repository where this diffing issue behavior could be observed?

**Comment by DaveSenn** [comment]
@nulltoken I try to find one.

**Comment by DaveSenn** [comment]
I’ve tested my code against a large amount of public repositories and wasn’t able to reproduce the exception. I also tried to create a test repository with a similar history… without success.
Do you have any ideas how I could reproduce the exception (on a different repository)?
Otherwise, I’ll try to debug the library code (as soon as I’ve time…)

**Comment by nulltoken** [comment]
>Otherwise, I’ll try to debug the library code (as soon as I’ve time…)

Considering the call stack, there's a huge chance the problem comes the native code. Any chance you could attach a debugger and help us identifying which part of libgit2 causes this? That would be the quickest way to get this fixed.

If native code debugging isn't your sweet spot, we thought of a potential alternative. There may be a way to help us get a grasp at the commits being compared, and maybe help us figuring out what could cause this error.

The code below dumps the content of a commit. For each object in the commit, it renders its Sha, type, size and mode. It also obfuscates the path of each object by hashing the path segments.

If that's fine by you, could you please run `DumpCommit()` against both the commits that are compared when the exception is thrown and paste the result here?

```csharp
private void DumpCommit(IRepository repo, Commit commit)
{
    var hasher = System.Security.Cryptography.SHA1.Create("System.Security.Cryptography.SHA1");
    Console.WriteLine("Commit {0}", commit.Sha);
    Console.WriteLine();
    DumpTree(repo, "/", commit.Tree, hasher);
    Console.WriteLine("--------");
}

private void DumpTree(IRepository repo, string prefix, Tree tree, System.Security.Cryptography.SHA1 hasher)
{
    foreach (var treeEntry in tree)
    {
        var path = prefix + Obfuscate(hasher, treeEntry.Name);
        var gitObject = treeEntry.Target;

        var meta = repo.ObjectDatabase.RetrieveObjectMetadata(gitObject.Id);
        Console.WriteLine("{0}\t{1}\t{2}\t{3}\t{4}", gitObject.Id, treeEntry.Mode, treeEntry.TargetType, meta.Size, path);

        if (treeEntry.TargetType == TreeEntryTargetType.Tree)
        {
            DumpTree(repo, path + "/", (Tree)gitObject, hasher);
        }
    }
}

private object Obfuscate(System.Security.Cryptography.SHA1 hasher, string path)
{
    byte[] data = hasher.ComputeHash(Encoding.UTF8.GetBytes(path));

    StringBuilder sBuilder = new StringBuilder();

    foreach (byte t in data)
    {
        sBuilder.Append(t.ToString("x2"));
    }

    return sBuilder.ToString();
}
```

**Comment by nulltoken** [comment]
Our amazing own @carlosmn crafted some additional debugging at the libgit2 level. This should output to the console some numbers used in a division that may be the one failing.

You'll find below the patch and a link to a specific build of the native binaries you should be a drop-in replacement.

 - Locate the `packages` in a development environment, within the `Lib` folder, drop the `NativeBinaries` one and replace it with the one from the linked file .zip.
 - Drop the `NativeBinaries` from your `Debug` project folder
 - Rebuild (this will force the replacement of the patched binaries

With this patched version, calling `Diff.Compare(Tree, Tree)` should output something in the console. (Note: I haven't been able to see in in the Visual Studio output. However, running in from the command line worked perfectly).

Besides dumping the content of the Commits (as proposed above), could you please run the `Diff.Compare()`  call that throws using those binaries and paste the result in here?

```diff
diff --git a/src/hashsig.c b/src/hashsig.c
index 0ddfed9..d1a2dd3 100644
--- a/src/hashsig.c
+++ b/src/hashsig.c
@@ -328,6 +328,8 @@ static int hashsig_heap_compare(const hashsig_heap *a, const hashsig_heap *b)
                }
        }
 
+       printf("%s: a->size %d, b->size %d\n", "hashsig_heap_compare", a->size, b->size);
+
        return HASHSIG_SCALE * (matches * 2) / (a->size + b->size);
 }
```

https://www.dropbox.com/s/lygat9arx3kqv3r/NativeBinaries_979.zip?dl=0

**Comment by DaveSenn** [comment]
Ok I run your code for the commit ‘41ee66a46’ and its parents ‘d2be7a544’ and ‘2 80fda46b’.
(Comparing ‘41ee66a46’ and ‘d2be7a544’ throws the exception)

The output of the test is here https://drive.google.com/file/d/0B0vEstrhwNYRWmNncUlTanZybms/view?usp=sharing
exeOut: The stdout stream
commitDump: Output generated by the DumpCommit method
hashes: The full hases of the failing commit

**Comment by nulltoken** [comment]
@DaveSenn Aweome! :sparkles: 

Could also provide us with the CommitDump() of `41ee66a460f10ff4fb1ffa520f75a4671cf596c6` and `d2be7a544958b45d3c4c24dfd35cd3eef7d85a8d`?

Regarding the content of the `exeOut.txt` file, it looks like the console dump between `41ee66a460f10ff4fb1ffa520f75a4671cf596c6` and `d2be7a544958b45d3c4c24dfd35cd3eef7d85a8d` has been interrupted (maybe by the exception being thrown?). Any chance you could wrap your `Diff.Compare` in a `try/catch` block and maybe add a at the end a `Thread.Sleep` delay to allow the buffers to flush?

Considering the fact that the comparison between `41ee66a460f10ff4fb1ffa520f75a4671cf596c6` and `280fda46b20b21149bb565bc2cf97058e24f763b` "looks" like no console output has been produced, there may also be a possibility that the `printf` calls have been appended to the first comparison (due to a potential race condition in the way `printf` and `Console.WriteLine` flush their buffers), would you have run the two comparison in a row. If that's the case, we'd be very grateful if you could run them one at a time and maybe provide us with two separate outputs.

Once again, thanks a lot for the great help!

**Comment by DaveSenn** [comment]
That's the least I can do. 
Thanks for your great work  Thanks you for doing such a great work writing this library.

Sorry for the useless output… I was under a lot of stress and didn’t checked it carefully enough.
1)	Hmm strange… the call to DumpCommit was wrapped in a tray catch block… I’ll try to fix this behavior. 
2)	I will change the code so that I can run each comparison separately.

I’ve planned to do the test now… but currently I’ve no connection to our Git server (VPN makes some trouble). I run the test tomorrow morning, when I’m back at work.

Btw. This was my test code (Exception should get caught)
  

```csharp
private void Dump(LibGit2Sharp.Repository gitRepository)
{
	var failingCommit = gitRepository.Lookup<Commit>("41ee66a460f10ff4fb1ffa520f75a4671cf596c6");
	var failingCommitParent1 = gitRepository.Lookup<Commit>("d2be7a544958b45d3c4c24dfd35cd3eef7d85a8d");
	var failingCommitParent2 = gitRepository.Lookup<Commit>("280fda46b20b21149bb565bc2cf97058e24f763b");

	try
	{
		Console.WriteLine("{0}{0}Compare '{1}'(child) and {2}(parent){0}{0}", Environment.NewLine, failingCommit.Sha, failingCommitParent1.Sha);
		var changes = gitRepository.Diff.Compare<TreeChanges>(failingCommitParent1.Tree, failingCommit.Tree);
	}
	catch (DivideByZeroException)
	{
	}
	Console.WriteLine("{0}{0}Compare '{1}'(child) and {2}(parent){0}{0}", Environment.NewLine, failingCommit.Sha, failingCommitParent2.Sha);
	var changes1 = gitRepository.Diff.Compare<TreeChanges>(failingCommitParent2.Tree, failingCommit.Tree);

	Console.WriteLine( Environment.NewLine );
	Console.WriteLine("Done");
	Console.ReadLine();            

	var dumper = new CommitDumper(@"C:\Temp\commitDump.txt");
	dumper.DumpCommit(gitRepository, failingCommit);
	dumper.DumpCommit(gitRepository, failingCommitParent1);
	dumper.DumpCommit(gitRepository, failingCommitParent2);
}

public class CommitDumper
{
	private readonly String _path;

	public CommitDumper( String path )
	{
		_path = path;
	}

	public void DumpCommit( IRepository repo, Commit commit )
	{
		var hasher = SHA1.Create( "System.Security.Cryptography.SHA1" );

		using ( var writer = new StreamWriter( _path, true ) )
		{
			writer.WriteLine( "Commit {0}", commit.Sha );
			writer.WriteLine();
			DumpTree( repo, "/", commit.Tree, hasher, writer );
			writer.WriteLine( "--------" );
			writer.WriteLine(Environment.NewLine + Environment.NewLine);
		}
	}

	private void DumpTree( IRepository repo, String prefix, IEnumerable<TreeEntry> tree, SHA1 hasher, TextWriter writer )
	{
		foreach ( var treeEntry in tree )
		{
			var path = prefix + Obfuscate(hasher, treeEntry.Name);
			var gitObject = treeEntry.Target;

			var meta = repo.ObjectDatabase.RetrieveObjectMetadata( gitObject.Id );
			writer.WriteLine("{0}\t{1}\t{2}\t{3}\t{4}", gitObject.Id, treeEntry.Mode, treeEntry.TargetType, meta.Size, path);

			if ( treeEntry.TargetType == TreeEntryTargetType.Tree )
				DumpTree( repo, path + "/", (Tree) gitObject, hasher, writer );
		}
	}

	private Object Obfuscate( HashAlgorithm hasher, String path )
	{
		var data = hasher.ComputeHash( Encoding.UTF8.GetBytes( path ) );

		var sBuilder = new StringBuilder();

		foreach ( var t in data )
			sBuilder.Append( t.ToString( "x2" ) );

		return sBuilder.ToString();
	}
}
```

**Comment by DaveSenn** [comment]
Done, I’ve run the tests again. I’ve run each diff separately to avoid problems with printf.

Interesting is, comparing '41ee66a460f10ff4fb1ffa520f75a4671cf596c6' (child) and 'd2be7a544958b45d3c4c24dfd35cd3eef7d85a8d' (parent1) created a lot of output.
But comparing '41ee66a460f10ff4fb1ffa520f75a4671cf596c6' (child) and 280fda46b20b21149bb565bc2cf97058e24f763b' (parent2) generated not a single line.

The output is here: https://github.com/DaveSenn/libgit2sharpTests (inc. source code).

**Comment by nulltoken** [comment]
Bingo! 

```
hashsig_heap_compare: a->size 0, b->size 0
```

/cc @carlosmn @ethomson

**Comment by nulltoken** [comment]
The size here is 0. Seems weird.

@DaveSenn Could you run the same program locally without obfuscating the paths and check the content of the files which sha is `e69de29bb2d1d6434b8b29ae775ad8c2e48c5391`?

FWIW, `e69de29bb2d1d6434b8b29ae775ad8c2e48c5391` is a well known git sha for an empty Blob.

```
e69de29bb2d1d6434b8b29ae775ad8c2e48c5391 NonExecutableFile Blob 00000  /3960ec4ca5fb5e5d8cdb2cc1c5121c003e426517/a8323a2a9b34ae7e0cbe8181ebcfe45bf2c5f3aa/381e3298b2b8f6caeb2208b57d805ada38402f0b/53279e5ffd6cdb33896306457149914415d0a02b/9c9b70e8e04ad8f3f6ff4d04f441e3571d083f7f
```

**Comment by nulltoken** [comment]
@DaveSenn In order to ease the troubleshooting, @carlosmn came up with a a slightly modified libgit2 patch which should output the sha of the blobs being compared. This should also take care of flushing the buffers from the native side.

```diff
 diff --git a/src/diff_tform.c b/src/diff_tform.c
index 9133a9b..dcdd085 100644
--- a/src/diff_tform.c
+++ b/src/diff_tform.c
@@ -556,6 +556,8 @@ static int similarity_measure(
        bool exact_match = FLAG_SET(opts, GIT_DIFF_FIND_EXACT_MATCH_ONLY);
        int error = 0;
        similarity_info a_info, b_info;
+       char id_a[GIT_OID_HEXSZ + 1] = {0};
+       char id_b[GIT_OID_HEXSZ + 1] = {0};
 
        *score = -1;
 
@@ -619,6 +621,9 @@ static int similarity_measure(
        /* calculate similarity provided that the metric choose to process
         * both the a and b files (some may not if file is too big, etc).
         */
+       git_oid_fmt(id_a, &a_file->id);
+       git_oid_fmt(id_b, &b_file->id);
+       printf("similarity_measure: %s against %s\n", id_a, id_b);
        if (cache[a_idx] && cache[b_idx])
                error = opts->metric->similarity(
                        score, cache[a_idx], cache[b_idx], opts->metric->payload);
diff --git a/src/hashsig.c b/src/hashsig.c
index 0ddfed9..715d635 100644
--- a/src/hashsig.c
+++ b/src/hashsig.c
@@ -328,6 +328,9 @@ static int hashsig_heap_compare(const hashsig_heap *a, const hashsig_heap *b)
                }
        }
 
+       printf("hashsig_heap_compare: a->size %d, b->size %d\n", a->size, b->size);
+       fflush(stdout);
+
        return HASHSIG_SCALE * (matches * 2) / (a->size + b->size);
 }
```

You can download another version of the NativeBinaries version with this patch here: https://www.dropbox.com/s/g2rmh2cca0yweo3/NativeBinaries_979_2.zip?dl=0

This should produce similar results to

```
similarity_measure: d68dd4031d2ad5b7a3829ad7df6635e27a7daa22 against 94ebaf900161394059478fd88aec30e59092a1d7
hashsig_heap_compare: a->size 4, b->size 4
similarity_measure: 3aac70f843f29afb5bee93055fa1862bf07f7932 against 94ebaf900161394059478fd88aec30e59092a1d7
hashsig_heap_compare: a->size 4, b->size 4
similarity_measure: d68dd4031d2ad5b7a3829ad7df6635e27a7daa22 against 3aac70f843f29afb5bee93055fa1862bf07f7932
hashsig_heap_compare: a->size 4, b->size 4
similarity_measure: 94ebaf900161394059478fd88aec30e59092a1d7 against 3aac70f843f29afb5bee93055fa1862bf07f7932
hashsig_heap_compare: a->size 4, b->size 4
similarity_measure: 94ebaf900161394059478fd88aec30e59092a1d7 against d68dd4031d2ad5b7a3829ad7df6635e27a7daa22
hashsig_heap_compare: a->size 4, b->size 4
similarity_measure: 3aac70f843f29afb5bee93055fa1862bf07f7932 against d68dd4031d2ad5b7a3829ad7df6635e27a7daa22
hashsig_heap_compare: a->size 4, b->size 4
```

The divide by zero exception happens when the sum of `a->size` and `b->size` is equal to `0`.

When that happens in your environment, we'd be very interested in getting some more insight about the content of the files being compared.

**Comment by DaveSenn** [comment]
Sorry for the late response, I was skiing this weekend.
I’ve run my test tool with the new version of the native binaries.
The output is in ‘output_02_03_15’
-	exeOut.txt The output generated from the native binaries
-	commitDump.txt The commit dump (filtered by id ‘e69de29bb2d1d6434b8b29ae775ad8c2e48c5391’)
-	compareCommits.txt A diff of the 3 commits

The new output is in the git repo here: https://github.com/DaveSenn/libgit2sharpTests 

I’ve also checked the content of the files being compared.
All of them where empty (all are text files), expect 9 php files which do not exist.

Sorry that I’ve to censor so much information. I know this makes the whole thing not easier, but I’m not allowed to publish any company related data.

**Comment by nulltoken** [comment]
@DaveSenn Success! 

>similarity_measure: e69de29bb2d1d6434b8b29ae775ad8c2e48c5391 against 139597f9cb07c5d48bed18984ec4747f4b4f3438
>hashsig_heap_compare: a->size 0, b->size 0

I then googled for `139597f9cb07c5d48bed18984ec4747f4b4f3438`, hoping that some other repo would contain this nearly empty file... and found it. A blob with only two newline characters bears this sha.

Thanks to your output, I've successfully reproduced the issue.

```csharp
[Fact]
public void CanFindSimilaritiesBetweenAnEmptyBlobAndANearlyEmptyOne()
{
    string repoPath = InitNewRepository();

    using (var repo = new Repository(repoPath))
    {
        var emptyBlob = repo.Lookup<Blob>("e69de29bb2d1d6434b8b29ae775ad8c2e48c5391");
        var td = new TreeDefinition()
            .Add("empty.txt", emptyBlob, Mode.NonExecutableFile);

        var treeWithEmpty = repo.ObjectDatabase.CreateTree(td);

        Blob nearlyEmptyBlob = OdbHelper.CreateBlob(repo, "\n\n");
        td = new TreeDefinition()
            .Add("nearly.txt", nearlyEmptyBlob, Mode.NonExecutableFile);

        var treeWithNearly = repo.ObjectDatabase.CreateTree(td);

        var changes = repo.Diff.Compare<TreeChanges>(treeWithEmpty, treeWithNearly);

        Assert.NotNull(changes);
    }
}
```

@ethomson @carlosmn Thoughts?

**Comment by rtwede** [comment]
We also encountered this same issue.  It appears it has already been understood so I'd be glad to try a fix if one becomes available.

**Comment by ethomson** [comment]
Fixed by https://github.com/libgit2/libgit2/pull/2945

**Comment by nulltoken** [comment]
>When a queue of commits was divided between 8 threads all running their commit diffs in parallel the libgit2 library incurred an AccessViolation exception, likely due to something in the diff portion of libgit2 that is not thread safe.

@rtwede Could you please log a separate issue for this?

**Comment by SinghVarun** [comment]
I am too facing the same issue. Now as the issue is fixed in libgit2 project, when do you folks anticipate incorporating this dll under libgit2sharp project.

**Comment by nulltoken** [comment]
The fix has been embedded into LibGit2Sharp vNext as part of #965 (which has been published as a NuGet pre-relase package `LibGit2Sharp 0.22.0-pre20150317225305`)
